tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -3095,110 +3091,206 @@ LogicalResult ConvertTFLFakeQuantOp::matchAndRewrite(
 
 void LegalizeTFL::runOnFunction() {
   OwningRewritePatternList patterns(&getContext());
-  populateLegalizeTFLPatterns(&getContext(), patterns);
+  populateLegalizeTFLPatterns(&getContext(), patterns, legalization_enable);
 
   auto func = getFunction();
   if (ApplyPatternsWithShapeResolution(func, std::move(patterns)).failed()) {
     signalPassFailure();
   }
 }
+
+template <typename TFLOpT, typename TosaOpT>
+class ConvertTFLUnaryOp : public OpConversionPattern<TFLOpT> {
+public:
+  using OpConversionPattern<TFLOpT>::OpConversionPattern;
+  using OpAdaptor = typename TFLOpT::Adaptor;
+  LogicalResult matchAndRewrite(
+    TFLOpT op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+      rewriter.replaceOpWithNewOp<TosaOpT>(
+         op, op.getType(), adaptor.getOperands()[0]);
+      return success();
+  }
+};
+
+template <typename TFLOpT, typename TosaOpT>
+class ConvertTFLBinaryOp : public OpConversionPattern<TFLOpT> {
+public:
+  using OpConversionPattern<TFLOpT>::OpConversionPattern;
+  using OpAdaptor = typename TFLOpT::Adaptor;
+  LogicalResult matchAndRewrite(
+    TFLOpT op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+      rewriter.replaceOpWithNewOp<TosaOpT>(
+         op, op.getType(), adaptor.getOperands()[0], adaptor.getOperands()[1]);
+      return success();
+  }
+};
+
 }  // namespace
 
 void populateLegalizeTFLPatterns(MLIRContext* ctx,
-                                 RewritePatternSet& patterns) {
-  // Add the generated patterns to the list.
+                                 RewritePatternSet& patterns,
+                                 std::unordered_map<std::string, bool> legalization_enable) {
+  #define DEF_PATTERN_INSERT(PAT) patterns.insert<Convert##PAT##Op>(ctx);
+  #define DEF_PATTERN_INSERT_UNARY(PAT) patterns.insert<ConvertTFLUnaryOp<TFL::PAT##Op, tosa::PAT##Op>>(ctx);
+  #define DEF_PATTERN_INSERT_BINARY(PAT) patterns.insert<ConvertTFLBinaryOp<TFL::PAT##Op, tosa::PAT##Op>>(ctx);
+
+  /*
```

Comment:
We use // for block comments.

The behavior should be:

// Include pattern if not in the disabled list.

Reply:
Done